connector/src/main/java/org/geysermc/connector/network/translators/inventory/updater/AnvilInventoryUpdater.java

```diff
@@ -0,0 +1,461 @@
+        Map<JavaEnchantment, Integer> combinedEnchantments = getEnchantments(session, input);
+        int cost = 0;
+        for (Map.Entry<JavaEnchantment, Integer> entry : getEnchantments(session, material).entrySet()) {
+            JavaEnchantment enchantment = entry.getKey();
+            EnchantmentData data = Registries.ENCHANTMENTS.get(enchantment);
+            if (data == null) {
+                GeyserConnector.getInstance().getLogger().debug("Java enchantment not in registry: " + enchantment);
+                continue;
+            }
+
+            boolean canApply = isEnchantedBook(session, input) || data.validItems().contains(input.getJavaId());
+            for (JavaEnchantment incompatible : data.incompatibleEnchantments()) {
+                if (combinedEnchantments.containsKey(incompatible)) {
+                    canApply = false;
+                    if (!bedrock) {
+                        cost++;
+                    }
+                }
+            }
+
+            if (canApply || (!bedrock && session.getGameMode() == GameMode.CREATIVE)) {
+                int currentLevel = combinedEnchantments.getOrDefault(enchantment, 0);
+                int newLevel = entry.getValue();
+                if (newLevel == currentLevel) {
+                    newLevel++;
+                }
+                newLevel = Math.max(currentLevel, newLevel);
+                if (newLevel > data.maxLevel()) {
+                    newLevel = data.maxLevel();
+                }
+                combinedEnchantments.put(enchantment, newLevel);
+
+                int rarityMultiplier = data.rarityMultiplier();
+                if (isEnchantedBook(session, material) && rarityMultiplier > 1) {
+                    rarityMultiplier /= 2;
+                }
+                if (bedrock) {
+                    if (newLevel > currentLevel) {
+                        hasCompatible = true;
+                    }
+                    if (enchantment == JavaEnchantment.IMPALING) {
+                        // Multiplier is halved on Bedrock for some reason
+                        rarityMultiplier /= 2;
+                    } else if (enchantment == JavaEnchantment.SWEEPING) {
+                        // Doesn't exist on Bedrock
+                        rarityMultiplier = 0;
+                    }
+                    cost += rarityMultiplier * (newLevel - currentLevel);
+                } else {
+                    hasCompatible = true;
+                    cost += rarityMultiplier * newLevel;
+                }
+            }
+        }
+
+        if (!hasCompatible) {
+            return -1;
+        }
+        return cost;
+    }
+
+    private Map<JavaEnchantment, Integer> getEnchantments(GeyserSession session, GeyserItemStack itemStack) {
+        if (itemStack.getNbt() == null) {
+            return Object2ObjectMaps.emptyMap();
+        }
+        Map<JavaEnchantment, Integer> enchantments = new EnumMap<>(JavaEnchantment.class);
```

Comment:
What you can probably do is swap this out for a Reference2IntMap. If I am correct, Reference Fastutil maps act like IdentityHashMaps in that it compares objects and not hash codes. So there should still be similar performance, and you don't need to box integers. https://github.com/CloudburstMC/fastutil/blob/master/maps/reference-maps/reference-int-maps/pom.xml

Reply:
If I'm looking at this correctly, Reference2IntOpenHashMap and IdentityHashMap both use `System.identityHashCode` whereas EnumMap just uses the enum value's ordinal. I guess that would make Reference2IntMap behave exactly like an Object2IntMap at least for enums, since you can't override hashcode.